design/coreos.md

```diff
@@ -0,0 +1,209 @@
+- Simplifies customization of Ironic Python Agent by shipping it as a
+  container.
+- Enables customization specific to CoreOS Installer.
+
+Compared to the live ISO approach this proposal:
+
+- Does not require disabling inspection and cleaning.
+- Allows additional customizations on top of what CoreOS Installer does,
+  e.g. a potential fix for the [HPE virtual media
+  issue](https://storyboard.openstack.org/#!/story/2008763).
+
+### Goals
+
+- Support Ironic Python Agent images based on CoreOS.
+- Support using `coreos-installer` for installation for CoreOS.
+- Use unmodified CoreOS images.
+
+Note: CoreOS IPA images will support installing any target images, not only
+CoreOS.
+
+### Non-Goals
+
+- Swap out Ironic for anything else.
+- Any customizations on top of the MVP.
+
+## Proposal
+
+There are two sub-features in this design proposal that are going to work
+together, but can be implemented sequentially:
+
+1. Enable standard Ironic workflow using a CoreOS based ramdisk.
+2. Support using `coreos-installer` to install CoreOS.
+
+A proof of concept of both features can be found here:
+<https://github.com/dtantsur/ironic-agent-image>.
+
+## Design Details
+
+Since we want to use unmodified CoreOS images (which makes a lot of sense from
+the maintainability perspective), we need an alternative way to start Ironic
+Python Agent on them. A canonical way to do is to to inject [Ignition
+configuration](https://coreos.github.io/ignition/examples/#start-services) that
+pulls a container and runs it with podman. [Example configuration
+template](https://github.com/dtantsur/ironic-agent-image/blob/main/ignition/ignition.json),
+[example service
+template](https://github.com/dtantsur/ironic-agent-image/blob/main/ignition/service).
+
+Note: Ignition is a CoreOS analog of cloud-init.
+
+We need to distinguish between two cases:
+
+1. PXE/iPXE booting
+2. ISO booting (with virtual media)
+
+CoreOS
+[publishes](https://getfedora.org/en/coreos/download?tab=metal_virtualized&stream=stable)
+both kinds of artifacts (and more), but the customization pattern is completely
+different.
+
+### PXE images
+
+CoreOS publishes 2 artifacts: a kernel, an initramfs and a root filesystem, all
+three are required for successful boot. The root filesystem and the Ignition
+configuration are specified as URLs via the kernel parameters.
+
+We will place the root filesystem and the generated Ignition configuration into
+the shared location where other HTTP artifacts are located. The Ironic's
+`pxe_append_params` will be updated with links to them.
+
+### ISO image
+
+We cannot modify kernel parameters on the ISO image without rebuilding it, thus
+we will use a different [customization
+procedure](https://coreos.github.io/coreos-installer/iso-embed-ignition/). On
+the Ironic start-up the downloaded ISO will be customized to include the
+generated Ignition configuration, including an Ironic Python Agent
+configuration file with the endpoint URLs (Ironic, Inspector, etc).
```

Comment:
This paragraph is confusing. It seems to be saying we would use a different approach from rebuilding the image, but then describes rebuilding the image.

Reply:
It's not rebuilding the image, it's injecting a file into a reserved space. Please see the link.